### ISSUE TITLE
False E0237 for __slots__ and collections.abc.Iterable and Container and subtypes using TypeVar specialization.

### ISSUE THREAD

**Issue by yumasheta** [PyCQA/pylint#5705]
### Bug description

Consider the following example:

```python
from collections import abc
from typing import TypeVar

_T = TypeVar("_T")

class Data(abc.Iterable[_T]):
    """The _data is in slots."""

    __slots__ = {"_data"}

    def __init__(self, elements):
        """Initialize with `Data` instances."""
        self._data = list(elements)  # E0237: Assigning to attribute '_data' not defined in class slots (assigning-non-slot

class Data(abc.Iterable[int]):
    """The _data is in slots."""

    __slots__ = {"_data"}

    def __init__(self, elements):
        """Initialize with `Data` instances."""
        self._data = list(elements)

class Data(abc.Container[_T]):
    """The _data is in slots."""

    __slots__ = {"_data"}

    def __init__(self, elements):
        """Initialize with `Data` instances."""
        self._data = list(elements)  # E0237: Assigning to attribute '_data' not defined in class slots (assigning-non-slot)

class Data(abc.Container[int]):
    """The _data is in slots."""

    __slots__ = {"_data"}

    def __init__(self, elements):
        """Initialize with `Data` instances."""
        self._data = list(elements)
```

False E0237 are issued for the valid code (except that the subtypes still have abstract methods).
The false error messages are only issued when the `abc` types are specialized with a `TypeVar`. Using explicit (in this case built-in) types raises no error.

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
E0237: Assigning to attribute '_data' not defined in class slots (assigning-non-slot)
E0237: Assigning to attribute '_data' not defined in class slots (assigning-non-slot)
```

### Expected behavior

No E0237 errors.

### Pylint version

```shell
pylint 2.13.0-dev0
astroid 2.9.3
Python 3.10.1 (main, Dec 18 2021, 23:53:45) [GCC 11.1.0]
```

### OS / Environment

linux 5.16.0-arch1-1

**Comment by Pierre-Sassoulas** [comment]
I can't reproduce on python 3.8 with the latest pylint.

**Comment by yumasheta** [comment]
interesting. I just checked on a fresh venv with python 3.10.5:
```
astroid           2.11.6
dill              0.3.5.1
isort             5.10.1
lazy-object-proxy 1.7.1
mccabe            0.7.0
pip               22.1.2
platformdirs      2.5.2
pylint            2.15.0.dev0 /tmp/pylint_test/pylint
setuptools        63.1.0
tomli             2.0.1
tomlkit           0.11.0
wrapt             1.14.1
```

And I can reproduce...
